components/script/dom/htmlformelement.rs

```diff
@@ -746,10 +747,28 @@ impl HTMLFormElement {
         let enctype = submitter.enctype();
         let method = submitter.method();
 
-        // Step 17-21
-        let target_attribute_value = submitter.target();
+        // Step 17
+        let target_attribute_value: Option<DOMString>;
+        let subject = submitter.form_owner().unwrap();
+        if submitter.is_submit_button() && submitter.target() != DOMString::new() {
+            target_attribute_value = Some(action);
+        } else {
+            target_attribute_value =
+                get_element_target(&subject.upcast::<HTMLElement>().upcast::<Element>());
+        }
+
+        // Step 18
+        let noopener = get_element_noopener(
+            &subject.upcast::<HTMLElement>().upcast::<Element>(),
```

Comment:
This should be`self.upcast::<Element>()`, based on my reading of the spec.

Reply:
Sounds good! I'll get to these in the morning.